pkg/apis/machine/v1beta1/machineset_webhook_test.go

```diff
@@ -0,0 +1,941 @@
+			mgr.GetWebhookServer().Register("/validate-machine-openshift-io-v1beta1-machineset-cp-delete", &webhook.Admission{Handler: createMachineSetMockHandler(true)})
+			mgr.GetWebhookServer().Register("/validate-machine-openshift-io-v1beta1-machineset-cp-update", &webhook.Admission{Handler: createMachineSetMockHandler(true)})
+
+			done := make(chan struct{})
+			stopped := make(chan struct{})
+			go func() {
+				gs.Expect(mgr.Start(done)).To(Succeed())
+				close(stopped)
+			}()
+			defer func() {
+				close(done)
+				<-stopped
+			}()
+
+			gs.Eventually(func() (bool, error) {
+				resp, err := insecureHTTPClient.Get(fmt.Sprintf("https://127.0.0.1:%d", testEnv.WebhookInstallOptions.LocalServingPort))
+				if err != nil {
+					return false, err
+				}
+				return resp.StatusCode == 404, nil
+			}).Should(BeTrue())
+
+			ms := &MachineSet{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "machineset-update-",
+					Namespace:    namespace.Name,
+				},
+				Spec: MachineSetSpec{
+					Template: MachineTemplateSpec{
+						Spec: MachineSpec{
+							ProviderSpec: ProviderSpec{
+								Value: tc.baseProviderSpecValue,
+							},
+						},
+					},
+				},
+			}
+			err = c.Create(ctx, ms)
+			gs.Expect(err).ToNot(HaveOccurred())
+			defer func() {
+				gs.Expect(c.Delete(ctx, ms)).To(Succeed())
+			}()
+
+			ms.Spec.Template.Spec.ProviderSpec.Value = tc.updatedProviderSpecValue()
+			err = c.Update(ctx, ms)
+			if tc.expectedError != "" {
+				gs.Expect(err).ToNot(BeNil())
+				gs.Expect(apierrors.ReasonForError(err)).To(BeEquivalentTo(tc.expectedError))
+			} else {
+				gs.Expect(err).To(BeNil())
+			}
+		})
+	}
+}
+
+func TestCPMachineSetDelete(t *testing.T) {
+	g := NewWithT(t)
+
+	// Override config getter
+	ctrl.GetConfig = func() (*rest.Config, error) {
+		return cfg, nil
+	}
+	defer func() {
+		ctrl.GetConfig = config.GetConfig
+	}()
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "machineset-cp-delete-test",
+		},
+	}
+	g.Expect(c.Create(ctx, namespace)).To(Succeed())
+	defer func() {
+		g.Expect(c.Delete(ctx, namespace)).To(Succeed())
+	}()
+
+	testCases := []struct {
+		name          string
+		expectedError string
+		objectMeta    ObjectMeta
+	}{
+		{
+			name:          "is not CP MachineSet",
+			expectedError: "",
+			objectMeta: ObjectMeta{
+				Labels: map[string]string{
+					"machine.openshift.io/cluster-api-machine-role": "worker",
+				},
+			},
+		},
+		{
+			name:          "is not CP MachineSet, no labels",
+			expectedError: "",
+			objectMeta:    ObjectMeta{},
+		},
+		{
+			name:          "is CP MachineSet",
+			expectedError: "Requested DELETE of Control Plane MachineSet Not Allowed.",
+			objectMeta: ObjectMeta{
+				Labels: map[string]string{
+					"machine.openshift.io/cluster-api-machine-role": "master",
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
```

Comment:
This is the test that fails most often without this patch.  With this patch, the tests run indefinitely until they hit the 10 minute testing timeout panic built into golang.